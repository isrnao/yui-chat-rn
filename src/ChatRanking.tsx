import type { Chat } from "./types";

type Props = {
    chatLog: Chat[];
};

function formatCountTime(ts: number): string {
    const d = new Date(ts);
    const w = ["日", "月", "火", "水", "木", "金", "土"];
    return `${d.getMonth() + 1}/${d.getDate()}(${w[d.getDay()]})${d.getHours().toString().padStart(2, "0")}:${d.getMinutes().toString().padStart(2, "0")}`;
}

export default function ChatRanking({ chatLog }: Props) {
    const ranking: { name: string, count: number, lastTime: number }[] = [];
    const map = new Map<string, { count: number, lastTime: number }>();
    chatLog.forEach(c => {
        if (!c.system && c.name) {
            const rec = map.get(c.name) ?? { count: 0, lastTime: 0 };
            rec.count += 1;
            rec.lastTime = Math.max(rec.lastTime, c.time);
            map.set(c.name, rec);
        }
    });
    map.forEach((v, k) => ranking.push({ name: k, ...v }));
    ranking.sort((a, b) => b.count - a.count || b.lastTime - a.lastTime);

    return (
        <div className="mt-4 w-full mx-auto text-[#444]">
            <div className="mb-2 text-sm font-bold">
                発言らんきんぐ
            </div>
            <div className="overflow-x-auto">
                <table className="w-full text-sm" style={{ fontFamily: "var(--font-yui)" }}>
                    <thead>
                        <tr className="text-[#444]">
                            <th className="py-1 px-2 font-bold whitespace-nowrap border-b border-[#d2b48c]">おなまえ</th>
                            <th className="py-1 px-2 font-bold whitespace-nowrap border-b border-[#d2b48c]">発言回数</th>
                            <th className="py-1 px-2 font-bold whitespace-nowrap border-b border-[#d2b48c]">最終発言時刻</th>
                            <th className="py-1 px-2 font-bold whitespace-nowrap border-b border-[#d2b48c]">ホスト情報</th>
                        </tr>
                    </thead>
                    <tbody>
                        {ranking.length === 0 && (
                            <tr>
                                <td colSpan={4} className="text-center text-gray-400 py-2">
                                    データなし
                                </td>
                            </tr>
                        )}
                        {ranking.map(({ name, count, lastTime }) => (
                            <tr key={name}>
                                <td className="py-1 px-2 border-b border-[#e9d7ba] text-center">{name}</td>
                                <td className="py-1 px-2 border-b border-[#e9d7ba] text-center">{count}</td>
                                <td className="py-1 px-2 border-b border-[#e9d7ba] text-center">{formatCountTime(lastTime)}</td>
                                <td className="py-1 px-2 border-b border-[#e9d7ba] text-xs text-gray-400 text-center">-</td>
                            </tr>
                        ))}
                    </tbody>
                </table>
            </div>
        </div>
    );
}
