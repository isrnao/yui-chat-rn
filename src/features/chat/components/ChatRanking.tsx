import { useChatRanking } from "@features/chat/hooks/useChatRanking";
import { formatCountTime } from "@shared/utils/format";
import type { Chat } from "@features/chat/types";

type Props = {
  chatLog: Chat[];
};

export default function ChatRanking({ chatLog }: Props) {
  const ranking = useChatRanking(chatLog);

  return (
    <div className="mt-4 w-full mx-auto text-[#444]">
      <div className="mb-2 text-sm font-bold">発言らんきんぐ</div>
      <div className="overflow-x-auto">
        <table
          className="w-full text-sm"
          style={{ fontFamily: "var(--font-yui)" }}
        >
          <thead>
            <tr className="text-[#444]">
              <th className="py-1 px-2 font-bold whitespace-nowrap border-b border-[#d2b48c]">
                おなまえ
              </th>
              <th className="py-1 px-2 font-bold whitespace-nowrap border-b border-[#d2b48c]">
                発言回数
              </th>
              <th className="py-1 px-2 font-bold whitespace-nowrap border-b border-[#d2b48c]">
                最終発言時刻
              </th>
              <th className="py-1 px-2 font-bold whitespace-nowrap border-b border-[#d2b48c]">
                ホスト情報
              </th>
            </tr>
          </thead>
          <tbody>
            {ranking.length === 0 && (
              <tr>
                <td colSpan={4} className="text-center text-gray-400 py-2">
                  データなし
                </td>
              </tr>
            )}
            {ranking.map(({ name, count, lastTime }) => (
              <tr key={name}>
                <td className="py-1 px-2 border-b border-[#e9d7ba] text-center">
                  {name}
                </td>
                <td className="py-1 px-2 border-b border-[#e9d7ba] text-center">
                  {count}
                </td>
                <td className="py-1 px-2 border-b border-[#e9d7ba] text-center">
                  {formatCountTime(lastTime)}
                </td>
                <td className="py-1 px-2 border-b border-[#e9d7ba] text-center">
                  {/* ホスト情報は必要に応じて追加 */}
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    </div>
  );
}
