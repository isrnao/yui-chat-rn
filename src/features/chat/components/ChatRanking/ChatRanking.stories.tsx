import type { Meta, StoryObj } from '@storybook/react';
import ChatRanking from './index';
import { sampleChatLog } from '../../../../storybook/mocks/chatSamples';

const meta = {
  component: ChatRanking,
  parameters: {
    layout: 'padded',
  },
  tags: ['autodocs'],
  args: {
    chatLog: sampleChatLog,
  },
  argTypes: {
    chatLog: {
      description: 'ランキング計算に利用するチャットログ',
    },
  },
} satisfies Meta<typeof ChatRanking>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {};

export const Empty: Story = {
  args: {
    chatLog: [],
  },
};
