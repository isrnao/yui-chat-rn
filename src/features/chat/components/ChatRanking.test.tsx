import { render, screen } from '@testing-library/react';
import ChatRanking from './ChatRanking';
import { vi } from 'vitest';

// formatCountTimeをモック（呼び出し確認もしやすい）
vi.mock('@shared/utils/format', () => ({
  formatCountTime: (time: string) => `formatted:${time}`,
}));

describe('<ChatRanking />', () => {
  it('チャット履歴が空の時、「データなし」を表示する', () => {
    render(<ChatRanking chatLog={[]} />);
    expect(screen.getByText(/データなし/)).toBeInTheDocument();
  });

  it('ランキングが正しく表示される', () => {
    // 仮のチャットログデータ
    const chatLog = [
      { name: 'みどり', text: 'hi', createdAt: '2024-06-01T12:00:00Z' },
      { name: 'みどり', text: 'hello', createdAt: '2024-06-01T12:01:00Z' },
      { name: 'ゆい', text: 'やっほー', createdAt: '2024-06-01T12:02:00Z' },
    ];

    render(<ChatRanking chatLog={chatLog} />);
    // 名前が表示される
    expect(screen.getByText('みどり')).toBeInTheDocument();
    expect(screen.getByText('ゆい')).toBeInTheDocument();
    // 発言回数
    expect(screen.getByText('2')).toBeInTheDocument();
    expect(screen.getByText('1')).toBeInTheDocument();
    // formatCountTimeの返り値（最終発言時刻カラム）
    expect(screen.getAllByText(/^formatted:/).length).toBeGreaterThan(0);
  });

  it('スナップショットが一致する', () => {
    const chatLog = [{ name: 'A', text: 'a', createdAt: '2024-06-15T10:00:00Z' }];
    const { container } = render(<ChatRanking chatLog={chatLog} />);
    expect(container).toMatchSnapshot();
  });
});
