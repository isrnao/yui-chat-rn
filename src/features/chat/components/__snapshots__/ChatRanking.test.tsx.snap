// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ChatRanking /> > スナップショットが一致する 1`] = `
<div>
  <div
    class="mt-4 w-full mx-auto text-[#444]"
  >
    <div
      class="mb-2 text-sm font-bold"
    >
      発言らんきんぐ
    </div>
    <div
      class="overflow-x-auto"
    >
      <table
        class="w-full text-sm"
        style="font-family: var(--font-yui);"
      >
        <thead>
          <tr
            class="text-[#444]"
          >
            <th
              class="py-1 px-2 font-bold whitespace-nowrap border-b border-[#d2b48c]"
            >
              おなまえ
            </th>
            <th
              class="py-1 px-2 font-bold whitespace-nowrap border-b border-[#d2b48c]"
            >
              発言回数
            </th>
            <th
              class="py-1 px-2 font-bold whitespace-nowrap border-b border-[#d2b48c]"
            >
              最終発言時刻
            </th>
            <th
              class="py-1 px-2 font-bold whitespace-nowrap border-b border-[#d2b48c]"
            >
              ホスト情報
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="py-1 px-2 border-b border-[#e9d7ba] text-center"
            >
              A
            </td>
            <td
              class="py-1 px-2 border-b border-[#e9d7ba] text-center"
            >
              1
            </td>
            <td
              class="py-1 px-2 border-b border-[#e9d7ba] text-center"
            >
              formatted:NaN
            </td>
            <td
              class="py-1 px-2 border-b border-[#e9d7ba] text-center"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
